### ISSUE TITLE
java.lang.StackOverflowError when codec does not have an encode method

### ISSUE THREAD

**Issue by TheVastyDeep** [elastic/logstash#13443]
logstash 7.15.2, installed using yum, run from the command line. The following configuration

    input { generator { count => 1 lines => [ '' ] } }
    output { udp { port  => 2055 codec => netflow host => "127.1.2.3" } }

results in 

    warning: thread "[main]>worker0" terminated with exception (report_on_exception is true):
    java.lang.StackOverflowError
        at usr.share.logstash.logstash_minus_core.lib.logstash.codecs.base.encode(/usr/share/logstash/logstash-core/lib/logstash/codecs/base.rb:48)
        at usr.share.logstash.logstash_minus_core.lib.logstash.codecs.base.multi_encode(/usr/share/logstash/logstash-core/lib/logstash/codecs/base.rb:64)
        at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1820)
        at usr.share.logstash.logstash_minus_core.lib.logstash.codecs.base.multi_encode(/usr/share/logstash/logstash-core/lib/logstash/codecs/base.rb:64)
        at usr.share.logstash.logstash_minus_core.lib.logstash.codecs.base.encode(/usr/share/logstash/logstash-core/lib/logstash/codecs/base.rb:48)
        at usr.share.logstash.logstash_minus_core.lib.logstash.codecs.base.multi_encode(/usr/share/logstash/logstash-core/lib/logstash/codecs/base.rb:64)
        at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1820)

because in [codec/base.rb](https://github.com/elastic/logstash/blob/3aa5741e687ab602169b4a716ff8820f622dd92b/logstash-core/lib/logstash/codecs/base.rb#L47) encode calls multi_encode, which calls encode.

Personally I think the netflow codec should be modified to add an encode method which throws an exception saying "You cannot use this codec on an output".

**Comment by kares** [comment]
Thanks for the report, indeed the base class assumed the codec to override `multi_encode` or `encode`.

> Personally I think the netflow codec should be modified to add an encode method which throws an exception saying "You cannot use this codec on an output".

Yes, this would be the best approach, ([e.g. a different codec not supporting `decode` raises](https://github.com/logstash-plugins/logstash-codec-dots/blob/v3.0.6/lib/logstash/codecs/dots.rb#L9)). 

It might be better to have the issue report at https://github.com/logstash-plugins/logstash-codec-netflow/issues so someone looking at the codec is aware of it.

**Comment by TheVastyDeep** [comment]
I have opened [196](https://github.com/logstash-plugins/logstash-codec-netflow/issues/196) against the netflow codec, so if the override is a requirement of the base filter this can be closed.

**Comment by kares** [comment]
yes due how codecs are handling either to re-def `encode` or `multi_encode` the *base.rb* isn't going to change anytime soon.
